
<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta charset="utf-8" />
    <title>Functions &#8212; V-gTK 0.1 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Building Rabies(RABV) resource" href="rabv-resource.html" />
    <link rel="prev" title="Prerequisite" href="prerequisite.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <div class="section" id="functions">
<h1>Functions<a class="headerlink" href="#functions" title="Permalink to this headline">¶</a></h1>
<div class="section" id="genbank-xml-downloader">
<h2>GenBank XML Downloader<a class="headerlink" href="#genbank-xml-downloader" title="Permalink to this headline">¶</a></h2>
<p>This script downloads and updates GenBank XML files for a given species.
It supports batch downloads, update mode, and configurable directories.</p>
<div class="section" id="usage">
<h3>Usage<a class="headerlink" href="#usage" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python script.py --taxid &lt;TAXID&gt; <span class="o">[</span>options<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="arguments">
<h3>Arguments<a class="headerlink" href="#arguments" title="Permalink to this headline">¶</a></h3>
<dl class="option">
<dt id="cmdoption-t">
<code class="sig-name descname">-t</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--taxid</code><code class="sig-prename descclassname"> &lt;TAXID&gt;</code><a class="headerlink" href="#cmdoption-t" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>(Required)</strong> Taxonomic ID of the species.
Example: <code class="docutils literal notranslate"><span class="pre">11292</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-o">
<code class="sig-name descname">-o</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--tmp_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#cmdoption-o" title="Permalink to this definition">¶</a></dt>
<dd><p>Temporary output directory where XML files are stored.
Default: <code class="docutils literal notranslate"><span class="pre">tmp</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-b">
<code class="sig-name descname">-b</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--batch_size</code><code class="sig-prename descclassname"> &lt;N&gt;</code><a class="headerlink" href="#cmdoption-b" title="Permalink to this definition">¶</a></dt>
<dd><p>Maximum number of XML files to download and merge into a single file.
Default: <code class="docutils literal notranslate"><span class="pre">100</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-update">
<code class="sig-name descname">--update</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-update" title="Permalink to this definition">¶</a></dt>
<dd><p>Run in <em>update mode</em> to download only new sequences.
Expects a TSV file with a column named <code class="docutils literal notranslate"><span class="pre">Accession</span> <span class="pre">Version</span></code>.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-u">
<code class="sig-name descname">-u</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--base_url</code><code class="sig-prename descclassname"> &lt;URL&gt;</code><a class="headerlink" href="#cmdoption-u" title="Permalink to this definition">¶</a></dt>
<dd><p>Base URL used to download the XML files.
Default: <code class="docutils literal notranslate"><span class="pre">https://eutils.ncbi.nlm.nih.gov/entrez/eutils/</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-e">
<code class="sig-name descname">-e</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--email</code><code class="sig-prename descclassname"> &lt;EMAIL&gt;</code><a class="headerlink" href="#cmdoption-e" title="Permalink to this definition">¶</a></dt>
<dd><p>Email ID for NCBI requests.
Default: <code class="docutils literal notranslate"><span class="pre">your_email&#64;example.com</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-s">
<code class="sig-name descname">-s</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--sleep_time</code><code class="sig-prename descclassname"> &lt;N&gt;</code><a class="headerlink" href="#cmdoption-s" title="Permalink to this definition">¶</a></dt>
<dd><p>Delay (in seconds) between fetch requests.
Default: <code class="docutils literal notranslate"><span class="pre">2</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-d">
<code class="sig-name descname">-d</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--base_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#cmdoption-d" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory where all GenBank XML files are stored.
Default: <code class="docutils literal notranslate"><span class="pre">GenBank-XML</span></code></p>
</dd></dl>

</div>
<div class="section" id="example">
<h3>Example<a class="headerlink" href="#example" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python script.py --taxid <span class="m">11292</span> --email myname@domain.com --batch_size <span class="m">200</span> --update
</pre></div>
</div>
</div>
</div>
<div class="section" id="genbank-xml-parser">
<h2>GenBank XML Parser<a class="headerlink" href="#genbank-xml-parser" title="Permalink to this headline">¶</a></h2>
<p>This command extracts information from GenBank XML files and writes a tab-separated table (TSV).</p>
<div class="section" id="id1">
<h3>Usage<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python GenBankParser.py --ref_list &lt;REFS_TSV&gt; <span class="o">[</span>options<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="id2">
<h3>Arguments<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h3>
<dl class="option">
<dt id="cmdoption-input-dir">
<code class="sig-name descname">-d</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--input_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#cmdoption-input-dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Input directory containing GenBank XML files.
Default: <code class="docutils literal notranslate"><span class="pre">tmp/GenBank-XML</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-base-dir">
<code class="sig-name descname">-b</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--base_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#cmdoption-base-dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Base working directory used by the extractor.
Default: <code class="docutils literal notranslate"><span class="pre">tmp</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-output-dir">
<code class="sig-name descname">-o</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--output_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#cmdoption-output-dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory where TSV tables are written.
Default: <code class="docutils literal notranslate"><span class="pre">GenBank-matrix</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-r">
<code class="sig-name descname">-r</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--ref_list</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-r" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>(Required)</strong> File containing a set of reference accessions.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-exclusion-list">
<code class="sig-name descname">-e</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--exclusion_list</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-exclusion-list" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional file with accessions to exclude from the output.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-is-segmented-virus">
<code class="sig-name descname">-s</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--is_segmented_virus</code><code class="sig-prename descclassname"> {Y,N}</code><a class="headerlink" href="#cmdoption-is-segmented-virus" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether the virus is segmented (<code class="docutils literal notranslate"><span class="pre">Y</span></code> or <code class="docutils literal notranslate"><span class="pre">N</span></code>).
Default: <code class="docutils literal notranslate"><span class="pre">N</span></code></p>
</dd></dl>

</div>
<div class="section" id="examples">
<h3>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h3>
<p>Basic run:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python GenBankParser.py <span class="se">\</span>
  --ref_list refs.tsv
</pre></div>
</div>
<p>Custom directories and segmented virus:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python GenBankParser.py <span class="se">\</span>
  --input_dir tmp/GenBank-XML <span class="se">\</span>
  --base_dir tmp <span class="se">\</span>
  --output_dir GenBank-matrix <span class="se">\</span>
  --ref_list refs.tsv <span class="se">\</span>
  --exclusion_list exclude.tsv <span class="se">\</span>
  --is_segmented_virus Y
</pre></div>
</div>
</div>
<div class="section" id="notes">
<h3>Notes<a class="headerlink" href="#notes" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Make sure your <code class="docutils literal notranslate"><span class="pre">--ref_list</span></code> file contains valid accession identifiers (one per line, or as expected by your parser).</p></li>
<li><p>When <code class="docutils literal notranslate"><span class="pre">--is_segmented_virus</span> <span class="pre">Y</span></code> is used, the output may include one row per segment depending on your parsing logic.</p></li>
</ul>
</div>
</div>
<div class="section" id="download-gff3-from-ncbi">
<h2>Download GFF3 from NCBI<a class="headerlink" href="#download-gff3-from-ncbi" title="Permalink to this headline">¶</a></h2>
<p>The script <code class="docutils literal notranslate"><span class="pre">DownloadGFF.py</span></code> downloads <strong>GFF3 files</strong> from NCBI using <em>efetch</em>.
It accepts one or more accession IDs and saves the results locally.</p>
<div class="section" id="id3">
<h3>Usage<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python DownloadGFF.py --accession_ids &lt;ACCESSION&gt; <span class="o">[</span>options<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="id4">
<h3>Arguments<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<dl class="option">
<dt id="cmdoption-id">
<code class="sig-name descname">-id</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--accession_ids</code><code class="sig-prename descclassname"> &lt;ID&gt;</code><a class="headerlink" href="#cmdoption-id" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>(Required)</strong> One or more NCBI accession IDs to download.
Example: <code class="docutils literal notranslate"><span class="pre">NC_002645</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id5">
<code class="sig-name descname">-b</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--base_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id5" title="Permalink to this definition">¶</a></dt>
<dd><p>Base working directory.
Default: <code class="docutils literal notranslate"><span class="pre">tmp</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id6">
<code class="sig-name descname">-o</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--output_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id6" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory where the downloaded GFF3 files are saved.
Default: <code class="docutils literal notranslate"><span class="pre">Gff</span></code></p>
</dd></dl>

</div>
<div class="section" id="id7">
<h3>Examples<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h3>
<p>Download a single accession:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python DownloadGFF.py --accession_ids NC_002645
</pre></div>
</div>
<p>Download into a custom directory:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python DownloadGFF.py <span class="se">\</span>
  --accession_ids NC_002645 <span class="se">\</span>
  --base_dir data <span class="se">\</span>
  --output_dir ncbi_gff
</pre></div>
</div>
</div>
</div>
<div class="section" id="validate-genbank-matrix">
<h2>Validate GenBank Matrix<a class="headerlink" href="#validate-genbank-matrix" title="Permalink to this headline">¶</a></h2>
<p>The script <code class="docutils literal notranslate"><span class="pre">ValidateMatrix.py</span></code> validates a GenBank matrix (<code class="docutils literal notranslate"><span class="pre">gB_matrix</span></code>) using
country, date, and host columns. It can download the NCBI taxonomy dump to
standardize names and produces a cleaned/validated TSV.</p>
<div class="section" id="id8">
<h3>Usage<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python ValidateMatrix.py <span class="o">[</span>options<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="id9">
<h3>Arguments<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h3>
<dl class="option">
<dt id="cmdoption-url">
<code class="sig-name descname">-u</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--url</code><code class="sig-prename descclassname"> &lt;URL&gt;</code><a class="headerlink" href="#cmdoption-url" title="Permalink to this definition">¶</a></dt>
<dd><p>URL to download the NCBI taxonomy dump (<code class="docutils literal notranslate"><span class="pre">taxdump.tar.gz</span></code>).
Default: <code class="docutils literal notranslate"><span class="pre">https://ftp.ncbi.nlm.nih.gov/pub/taxonomy/taxdump.tar.gz</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-taxa-path">
<code class="sig-name descname">-t</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--taxa_path</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#cmdoption-taxa-path" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory to store unpacked taxonomy files (names, nodes, etc.).
Default: <code class="docutils literal notranslate"><span class="pre">Taxa</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id10">
<code class="sig-name descname">-b</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--base_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id10" title="Permalink to this definition">¶</a></dt>
<dd><p>Base working directory.
Default: <code class="docutils literal notranslate"><span class="pre">tmp-test</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id11">
<code class="sig-name descname">-o</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--output_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id11" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory for validation results.
Default: <code class="docutils literal notranslate"><span class="pre">Validate-matrix</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-g">
<code class="sig-name descname">-g</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--gb_matrix</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-g" title="Permalink to this definition">¶</a></dt>
<dd><p>Input GenBank matrix TSV file.
Default: <code class="docutils literal notranslate"><span class="pre">tmp-test/GenBank-matrix/gB_matrix_raw.tsv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-c">
<code class="sig-name descname">-c</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--country</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-c" title="Permalink to this definition">¶</a></dt>
<dd><p>M49 country reference CSV used for normalization/validation.
Default: <code class="docutils literal notranslate"><span class="pre">assets/m49_country.csv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-a">
<code class="sig-name descname">-a</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--assets</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#cmdoption-a" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory containing auxiliary assets (e.g., reference CSVs).
Default: <code class="docutils literal notranslate"><span class="pre">assets/</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-no-debug">
<code class="sig-name descname">--no-debug</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-no-debug" title="Permalink to this definition">¶</a></dt>
<dd><p>Disable verbose/fuzzy debug prints.
(Debug logging is <strong>enabled by default</strong>; this flag turns it off.)</p>
</dd></dl>

</div>
<div class="section" id="id12">
<h3>Examples<a class="headerlink" href="#id12" title="Permalink to this headline">¶</a></h3>
<p>Basic validation with defaults:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python ValidateMatrix.py
</pre></div>
</div>
<p>Specify custom locations for inputs/outputs:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python ValidateMatrix.py <span class="se">\</span>
  --gb_matrix data/GenBank-matrix/gB_matrix_raw.tsv <span class="se">\</span>
  --country assets/m49_country.csv <span class="se">\</span>
  --base_dir work <span class="se">\</span>
  --output_dir Validate-matrix
</pre></div>
</div>
<p>Run quietly (no fuzzy debug prints):</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python ValidateMatrix.py --no-debug
</pre></div>
</div>
</div>
<div class="section" id="id13">
<h3>Notes<a class="headerlink" href="#id13" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>By default, the script enables debug messages; use <code class="docutils literal notranslate"><span class="pre">--no-debug</span></code> to suppress them.</p></li>
<li><p>Ensure the input matrix contains the expected columns (country, collection date,
host) so validation rules can be applied effectively.</p></li>
<li><p>The taxonomy dump will be downloaded to <code class="docutils literal notranslate"><span class="pre">--taxa_path</span></code> if not present.</p></li>
</ul>
</div>
</div>
<div class="section" id="add-missing-data-to-genbank-matrix">
<h2>Add Missing Data to GenBank Matrix<a class="headerlink" href="#add-missing-data-to-genbank-matrix" title="Permalink to this headline">¶</a></h2>
<p>The script <code class="docutils literal notranslate"><span class="pre">AddMissingData.py</span></code> processes a GenBank matrix to handle missing
values and apply bulk replacements. It can use a <em>fillup file</em> to insert
additional information and a <em>bulk file</em> for systematic replacements.</p>
<div class="section" id="id14">
<h3>Usage<a class="headerlink" href="#id14" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python AddMissingData.py <span class="o">[</span>options<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="id15">
<h3>Arguments<a class="headerlink" href="#id15" title="Permalink to this headline">¶</a></h3>
<dl class="option">
<dt id="cmdoption-tmp-dir">
<code class="sig-name descname">-d</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--tmp_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#cmdoption-tmp-dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory where the validated GenBank matrix is saved.
Recommended: <code class="docutils literal notranslate"><span class="pre">tmp/AddMissingData</span></code>
Default: <code class="docutils literal notranslate"><span class="pre">tmp/AddMissingData</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-gb-matrix">
<code class="sig-name descname">-g</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--gb_matrix</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-gb-matrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Input GenBank matrix TSV generated by <code class="docutils literal notranslate"><span class="pre">genbank_to_tsv.py</span></code>.
Default: <code class="docutils literal notranslate"><span class="pre">tmp/GenBank-matrix/gB_matrix_raw.tsv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-f">
<code class="sig-name descname">-f</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--fillup_file</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-f" title="Permalink to this definition">¶</a></dt>
<dd><p>TSV file containing additional metadata to fill missing values.
The file <strong>must</strong> contain a primary accession column; other columns are optional.
Default: <code class="docutils literal notranslate"><span class="pre">None</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-bulk-file">
<code class="sig-name descname">-b</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--bulk_file</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-bulk-file" title="Permalink to this definition">¶</a></dt>
<dd><p>TSV file specifying bulk replacements with <code class="docutils literal notranslate"><span class="pre">host</span></code> and <code class="docutils literal notranslate"><span class="pre">replaced_by</span></code> columns.
Default: <code class="docutils literal notranslate"><span class="pre">None</span></code></p>
</dd></dl>

</div>
<div class="section" id="id16">
<h3>Examples<a class="headerlink" href="#id16" title="Permalink to this headline">¶</a></h3>
<p>Run with default paths:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python AddMissingData.py
</pre></div>
</div>
<p>Provide a fillup file to supplement missing metadata:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python AddMissingData.py <span class="se">\</span>
  --fillup_file data/fillup.tsv
</pre></div>
</div>
<p>Apply bulk replacements in addition to fillups:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python AddMissingData.py <span class="se">\</span>
  --gb_matrix data/gB_matrix_raw.tsv <span class="se">\</span>
  --fillup_file data/fillup.tsv <span class="se">\</span>
  --bulk_file data/bulk_replace.tsv <span class="se">\</span>
  --tmp_dir tmp/AddMissingData
</pre></div>
</div>
</div>
<div class="section" id="id17">
<h3>Notes<a class="headerlink" href="#id17" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>The <strong>fillup file</strong> helps enrich the GenBank matrix with missing metadata.</p></li>
<li><p>The <strong>bulk file</strong> allows replacing multiple values systematically.</p></li>
<li><p>Output is written into the directory specified by <code class="docutils literal notranslate"><span class="pre">--tmp_dir</span></code>.</p></li>
</ul>
</div>
</div>
<div class="section" id="filter-and-extract-sequences">
<h2>Filter and Extract Sequences<a class="headerlink" href="#filter-and-extract-sequences" title="Permalink to this headline">¶</a></h2>
<p>The script <code class="docutils literal notranslate"><span class="pre">FilterAndExtractSequence.py</span></code> filters sequences from a GenBank matrix
and prepares them for downstream BLAST alignment. It supports filtering by sequence
length, ambiguity, GenBank divisions, and reference sets.</p>
<div class="section" id="id18">
<h3>Usage<a class="headerlink" href="#id18" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python FilterAndExtractSequence.py --genbank_matrix &lt;FILE&gt; --ref_file &lt;FILE&gt; <span class="o">[</span>options<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="id19">
<h3>Arguments<a class="headerlink" href="#id19" title="Permalink to this headline">¶</a></h3>
<dl class="option">
<dt id="cmdoption-genbank-matrix">
<code class="sig-name descname">-g</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--genbank_matrix</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-genbank-matrix" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>(Required)</strong> Input GenBank matrix file.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-sf">
<code class="sig-name descname">-sf</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--sequence_file</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-sf" title="Permalink to this definition">¶</a></dt>
<dd><p>Input FASTA file containing sequences.
Default: <code class="docutils literal notranslate"><span class="pre">tmp/GenBank-matrix/sequences.fa</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-genbank-matrix-filtered">
<code class="sig-name descname">-f</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--genbank_matrix_filtered</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#cmdoption-genbank-matrix-filtered" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory where the filtered GenBank matrix will be stored.
Default: <code class="docutils literal notranslate"><span class="pre">tmp/GenBank-matrix</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-ref-file">
<code class="sig-name descname">-r</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--ref_file</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-ref-file" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>(Required)</strong> Text file containing a list of reference sequence accessions.</p>
</dd></dl>

<dl class="option">
<dt id="id20">
<code class="sig-name descname">-b</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--base_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id20" title="Permalink to this definition">¶</a></dt>
<dd><p>Base working directory.
Default: <code class="docutils literal notranslate"><span class="pre">tmp</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id21">
<code class="sig-name descname">-o</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--output_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id21" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory where processed sequences will be written.
Default: <code class="docutils literal notranslate"><span class="pre">Sequences</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-l">
<code class="sig-name descname">-l</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--total_length</code><code class="sig-prename descclassname"> &lt;N&gt;</code><a class="headerlink" href="#cmdoption-l" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum sequence length (including ambiguous characters).
Default: <code class="docutils literal notranslate"><span class="pre">1</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-n">
<code class="sig-name descname">-n</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--real_length</code><code class="sig-prename descclassname"> &lt;N&gt;</code><a class="headerlink" href="#cmdoption-n" title="Permalink to this definition">¶</a></dt>
<dd><p>Minimum sequence length excluding ambiguous <code class="docutils literal notranslate"><span class="pre">N</span></code> bases.
Default: <code class="docutils literal notranslate"><span class="pre">1</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-prop-ambigious-data">
<code class="sig-name descname">-a</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--prop_ambigious_data</code><code class="sig-prename descclassname"> &lt;VALS&gt;</code><a class="headerlink" href="#cmdoption-prop-ambigious-data" title="Permalink to this definition">¶</a></dt>
<dd><p>Proportion(s) of ambiguous data to exclude.
Accepts one or more values.
Default: <code class="docutils literal notranslate"><span class="pre">None</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-v">
<code class="sig-name descname">-v</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--segmented_virus</code><code class="sig-prename descclassname"> &lt;Y/N&gt;</code><a class="headerlink" href="#cmdoption-v" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether the dataset is from a segmented virus.
Default: <code class="docutils literal notranslate"><span class="pre">N</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-genbank-division">
<code class="sig-name descname">-d</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--genbank_division</code><code class="sig-prename descclassname"> &lt;LIST&gt;</code><a class="headerlink" href="#cmdoption-genbank-division" title="Permalink to this definition">¶</a></dt>
<dd><p>GenBank divisions to exclude from analysis.
Default: <code class="docutils literal notranslate"><span class="pre">[&quot;VRL&quot;,</span> <span class="pre">&quot;PAT&quot;,</span> <span class="pre">&quot;SYN&quot;,</span> <span class="pre">&quot;ENV&quot;]</span></code></p>
<p>Common divisions include:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">VRL</span></code> — Viral sequences</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ENV</span></code> — Environmental sequences</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">PAT</span></code> — Patented sequences</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">SYN</span></code> — Synthetic/chimeric sequences</p></li>
</ul>
</dd></dl>

<dl class="option">
<dt id="cmdoption-vd">
<code class="sig-name descname">-vd</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--valid_divisions</code><code class="sig-prename descclassname"> &lt;LIST&gt;</code><a class="headerlink" href="#cmdoption-vd" title="Permalink to this definition">¶</a></dt>
<dd><p>GenBank divisions to <strong>include</strong> for analysis.
Default: <code class="docutils literal notranslate"><span class="pre">[&quot;VRL&quot;,</span> <span class="pre">&quot;ENV&quot;]</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-seq-type">
<code class="sig-name descname">-s</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--seq_type</code><code class="sig-prename descclassname"> &lt;TYPE&gt;</code><a class="headerlink" href="#cmdoption-seq-type" title="Permalink to this definition">¶</a></dt>
<dd><p>Sequence type (e.g., nucleotide, protein).
Default: <code class="docutils literal notranslate"><span class="pre">None</span></code></p>
</dd></dl>

</div>
<div class="section" id="id22">
<h3>Examples<a class="headerlink" href="#id22" title="Permalink to this headline">¶</a></h3>
<p>Filter sequences longer than 5000 bases:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python FilterAndExtractSequence.py <span class="se">\</span>
  --genbank_matrix data/gB_matrix.tsv <span class="se">\</span>
  --ref_file refs.txt <span class="se">\</span>
  --total_length <span class="m">5000</span>
</pre></div>
</div>
<p>Exclude environmental and patented sequences:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python FilterAndExtractSequence.py <span class="se">\</span>
  --genbank_matrix data/gB_matrix.tsv <span class="se">\</span>
  --ref_file refs.txt <span class="se">\</span>
  --genbank_division VRL PAT SYN ENV
</pre></div>
</div>
<p>Filter by valid divisions and only keep nucleotide sequences:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python FilterAndExtractSequence.py <span class="se">\</span>
  --genbank_matrix data/gB_matrix.tsv <span class="se">\</span>
  --ref_file refs.txt <span class="se">\</span>
  --valid_divisions VRL <span class="se">\</span>
  --seq_type nucleotide
</pre></div>
</div>
</div>
<div class="section" id="id23">
<h3>Notes<a class="headerlink" href="#id23" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Both <code class="docutils literal notranslate"><span class="pre">--genbank_division</span></code> and <code class="docutils literal notranslate"><span class="pre">--valid_divisions</span></code> can be provided as lists
(space-separated).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--total_length</span></code> and <code class="docutils literal notranslate"><span class="pre">--real_length</span></code> help enforce sequence quality.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--prop_ambigious_data</span></code> allows filtering sequences with excessive ambiguity.</p></li>
</ul>
</div>
</div>
<div class="section" id="blast-alignment">
<h2>BLAST Alignment<a class="headerlink" href="#blast-alignment" title="Permalink to this headline">¶</a></h2>
<p>The script <code class="docutils literal notranslate"><span class="pre">BlastAlignment.py</span></code> performs BLAST alignments of query sequences
against a given set of reference sequences. It can create a BLAST database
from the provided reference FASTA file and produces a table of top hits.</p>
<div class="section" id="id24">
<h3>Usage<a class="headerlink" href="#id24" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python BlastAlignment.py --master_acc &lt;ACCESSION&gt; <span class="o">[</span>options<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="id25">
<h3>Arguments<a class="headerlink" href="#id25" title="Permalink to this headline">¶</a></h3>
<dl class="option">
<dt id="cmdoption-q">
<code class="sig-name descname">-q</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--query_fa</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-q" title="Permalink to this definition">¶</a></dt>
<dd><p>Query FASTA file containing sequences to align.
Default: <code class="docutils literal notranslate"><span class="pre">tmp/Sequences/query_seq.fa</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-ref-fa">
<code class="sig-name descname">-r</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--ref_fa</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-ref-fa" title="Permalink to this definition">¶</a></dt>
<dd><p>Reference FASTA file used to create a BLAST database.
(This file will be indexed for BLAST searches.)
Default: <code class="docutils literal notranslate"><span class="pre">tmp/Sequences/ref_seq.fa</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id26">
<code class="sig-name descname">-b</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--base_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id26" title="Permalink to this definition">¶</a></dt>
<dd><p>Base working directory.
Default: <code class="docutils literal notranslate"><span class="pre">tmp</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id27">
<code class="sig-name descname">-t</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--output_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id27" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory for BLAST results.
Default: <code class="docutils literal notranslate"><span class="pre">Blast</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-output-file">
<code class="sig-name descname">-o</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--output_file</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-output-file" title="Permalink to this definition">¶</a></dt>
<dd><p>Output file name for BLAST top hits.
Default: <code class="docutils literal notranslate"><span class="pre">query_tophits.tsv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id28">
<code class="sig-name descname">-s</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--is_segmented_virus</code><code class="sig-prename descclassname"> {Y,N}</code><a class="headerlink" href="#id28" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether the virus is segmented (<code class="docutils literal notranslate"><span class="pre">Y</span></code> or <code class="docutils literal notranslate"><span class="pre">N</span></code>).
Default: <code class="docutils literal notranslate"><span class="pre">N</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-segment-file">
<code class="sig-name descname">-f</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--segment_file</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-segment-file" title="Permalink to this definition">¶</a></dt>
<dd><p>File containing information about viral segments.
Required only when <code class="docutils literal notranslate"><span class="pre">--is_segmented_virus</span> <span class="pre">Y</span></code>.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-m">
<code class="sig-name descname">-m</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--master_acc</code><code class="sig-prename descclassname"> &lt;ACCESSION&gt;</code><a class="headerlink" href="#cmdoption-m" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>(Required)</strong> Master reference accession.
Example: Rabies virus uses <code class="docutils literal notranslate"><span class="pre">NC_001542</span></code>.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-is-update">
<code class="sig-name descname">-u</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--is_update</code><code class="sig-prename descclassname"> {Y,N}</code><a class="headerlink" href="#cmdoption-is-update" title="Permalink to this definition">¶</a></dt>
<dd><p>If set to <code class="docutils literal notranslate"><span class="pre">Y</span></code>, skips BLAST on sequences that were already aligned,
useful for incremental updates.
Default: <code class="docutils literal notranslate"><span class="pre">N</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-k">
<code class="sig-name descname">-k</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--keep_blast_tmp_dir</code><code class="sig-prename descclassname"> {Y,N}</code><a class="headerlink" href="#cmdoption-k" title="Permalink to this definition">¶</a></dt>
<dd><p>Whether to retain the temporary BLAST working directory for debugging.
Default: <code class="docutils literal notranslate"><span class="pre">N</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id29">
<code class="sig-name descname">-g</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--gb_matrix</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#id29" title="Permalink to this definition">¶</a></dt>
<dd><p>Input GenBank matrix file.
Default: <code class="docutils literal notranslate"><span class="pre">tmp/GenBank-matrix/gB_matrix_raw.tsv</span></code></p>
</dd></dl>

</div>
<div class="section" id="id30">
<h3>Examples<a class="headerlink" href="#id30" title="Permalink to this headline">¶</a></h3>
<p>Run a basic BLAST alignment:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python BlastAlignment.py <span class="se">\</span>
  --master_acc NC_001542 <span class="se">\</span>
  --query_fa data/query.fa <span class="se">\</span>
  --ref_fa data/ref.fa
</pre></div>
</div>
<p>Run in update mode to skip previously aligned sequences:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python BlastAlignment.py <span class="se">\</span>
  --master_acc NC_001542 <span class="se">\</span>
  --query_fa new_sequences.fa <span class="se">\</span>
  --is_update Y
</pre></div>
</div>
<p>Debugging with retained BLAST temp directory:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python BlastAlignment.py <span class="se">\</span>
  --master_acc NC_001542 <span class="se">\</span>
  --keep_blast_tmp_dir Y
</pre></div>
</div>
</div>
<div class="section" id="id31">
<h3>Notes<a class="headerlink" href="#id31" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>The <strong>reference FASTA file</strong> is automatically converted into a BLAST database.</p></li>
<li><p>The <strong>master accession</strong> ensures consistency across analyses (e.g., Rabies virus = <code class="docutils literal notranslate"><span class="pre">NC_001542</span></code>).</p></li>
<li><p>Use <code class="docutils literal notranslate"><span class="pre">--is_update</span> <span class="pre">Y</span></code> for incremental analyses to save runtime.</p></li>
<li><p>Set <code class="docutils literal notranslate"><span class="pre">--keep_blast_tmp_dir</span> <span class="pre">Y</span></code> only when you need to inspect raw BLAST output for debugging.</p></li>
</ul>
</div>
</div>
<div class="section" id="nextalign-alignment">
<h2>Nextalign Alignment<a class="headerlink" href="#nextalign-alignment" title="Permalink to this headline">¶</a></h2>
<p>The script <code class="docutils literal notranslate"><span class="pre">NextalignAlignment.py</span></code> runs <strong>Nextalign</strong> for each query sequence,
aligning it to a master reference. It can optionally reuse a precomputed
reference alignment instead of generating one on the fly.</p>
<div class="section" id="id32">
<h3>Usage<a class="headerlink" href="#id32" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python NextalignAlignment.py --master_ref &lt;ACCESSION&gt; <span class="o">[</span>options<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="id33">
<h3>Arguments<a class="headerlink" href="#id33" title="Permalink to this headline">¶</a></h3>
<dl class="option">
<dt id="id34">
<code class="sig-name descname">-g</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--gB_matrix</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#id34" title="Permalink to this definition">¶</a></dt>
<dd><p>GenBank matrix (metadata) TSV file.
Default: <code class="docutils literal notranslate"><span class="pre">tmp/GenBank-matrix/gB_matrix_raw.tsv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-query-dir">
<code class="sig-name descname">-q</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--query_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#cmdoption-query-dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory containing grouped query FASTA files.
Default: <code class="docutils literal notranslate"><span class="pre">tmp/Blast/grouped_fasta</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-ref-dir">
<code class="sig-name descname">-r</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--ref_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#cmdoption-ref-dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory containing reference FASTA files.
Default: <code class="docutils literal notranslate"><span class="pre">tmp/Blast/ref_seqs</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-ref-fa-file">
<code class="sig-name descname">-f</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--ref_fa_file</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-ref-fa-file" title="Permalink to this definition">¶</a></dt>
<dd><p>Combined reference FASTA file.
This file is used to perform Nextalign against the master reference sequence.
Default: <code class="docutils literal notranslate"><span class="pre">tmp/Sequences/ref_seq.fa</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-ms">
<code class="sig-name descname">-ms</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--master_seq_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#cmdoption-ms" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory where the master sequence (derived from the master accession) is stored.
Default: <code class="docutils literal notranslate"><span class="pre">tmp/Blast/master_seq</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id35">
<code class="sig-name descname">-t</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--tmp_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id35" title="Permalink to this definition">¶</a></dt>
<dd><p>Temporary working directory.
Default: <code class="docutils literal notranslate"><span class="pre">tmp</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-master-ref">
<code class="sig-name descname">-m</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--master_ref</code><code class="sig-prename descclassname"> &lt;ACCESSION&gt;</code><a class="headerlink" href="#cmdoption-master-ref" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>(Required)</strong> Master reference accession (e.g., RefSeq).
Example (Rabies virus): <code class="docutils literal notranslate"><span class="pre">NC_001542</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-nextalign-dir">
<code class="sig-name descname">-n</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--nextalign_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#cmdoption-nextalign-dir" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory for Nextalign results.
Default: <code class="docutils literal notranslate"><span class="pre">Nextalign</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-ra">
<code class="sig-name descname">-ra</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--ref_alignment_file</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-ra" title="Permalink to this definition">¶</a></dt>
<dd><p>Optional <strong>reference alignment</strong> file to use instead of having Nextalign
align the reference to the master on the fly.</p>
</dd></dl>

</div>
<div class="section" id="id36">
<h3>Examples<a class="headerlink" href="#id36" title="Permalink to this headline">¶</a></h3>
<p>Run Nextalign using the default directories:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python NextalignAlignment.py <span class="se">\</span>
  --master_ref NC_001542
</pre></div>
</div>
<p>Specify a precomputed reference alignment file:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python NextalignAlignment.py <span class="se">\</span>
  --master_ref NC_001542 <span class="se">\</span>
  --ref_alignment_file data/reference_alignment.fasta
</pre></div>
</div>
<p>Use custom locations for queries and references:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python NextalignAlignment.py <span class="se">\</span>
  --master_ref NC_001542 <span class="se">\</span>
  --query_dir work/Blast/grouped_fasta <span class="se">\</span>
  --ref_dir work/Blast/ref_seqs <span class="se">\</span>
  --ref_fa_file work/Sequences/ref_seq.fa <span class="se">\</span>
  --nextalign_dir results/Nextalign
</pre></div>
</div>
</div>
<div class="section" id="id37">
<h3>Notes<a class="headerlink" href="#id37" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Ensure <strong>Nextalign</strong> is installed and available on your PATH before running the script.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--ref_alignment_file</span></code> is optional; if omitted, the script will let Nextalign
align reference sequences against the master reference automatically.</p></li>
<li><p>The <strong>master reference accession</strong> should correspond to the canonical reference
genome used for your analyses (e.g., Rabies virus = <code class="docutils literal notranslate"><span class="pre">NC_001542</span></code>).</p></li>
</ul>
</div>
</div>
<div class="section" id="pad-alignment">
<h2>Pad Alignment<a class="headerlink" href="#pad-alignment" title="Permalink to this headline">¶</a></h2>
<p>The script <code class="docutils literal notranslate"><span class="pre">PadAlignment.py</span></code> inserts gaps from a <strong>master alignment</strong>
into corresponding subalignments (e.g., from Nextalign output).
It ensures subalignments are padded consistently with the reference alignment
and merges them into a final alignment file.</p>
<div class="section" id="id38">
<h3>Usage<a class="headerlink" href="#id38" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python PadAlignment.py --reference_alignment &lt;FILE&gt; <span class="o">[</span>options<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="id39">
<h3>Arguments<a class="headerlink" href="#id39" title="Permalink to this headline">¶</a></h3>
<dl class="option">
<dt id="cmdoption-reference-alignment">
<code class="sig-name descname">-r</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--reference_alignment</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-reference-alignment" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>(Required)</strong> Master alignment file in FASTA format.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-i">
<code class="sig-name descname">-i</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--input_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#cmdoption-i" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory containing subalignment files (e.g., Nextalign output).
Default: <code class="docutils literal notranslate"><span class="pre">tmp/Nextalign/query_aln</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id40">
<code class="sig-name descname">-d</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--base_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id40" title="Permalink to this definition">¶</a></dt>
<dd><p>Base working directory.
Default: <code class="docutils literal notranslate"><span class="pre">tmp</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id41">
<code class="sig-name descname">-o</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--output_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id41" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory to save padded subalignments and merged files.
Default: <code class="docutils literal notranslate"><span class="pre">Pad-alignment</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-keep-intermediate-files">
<code class="sig-name descname">--keep_intermediate_files</code><code class="sig-prename descclassname"></code><a class="headerlink" href="#cmdoption-keep-intermediate-files" title="Permalink to this definition">¶</a></dt>
<dd><p>Retain intermediate padded subalignment files.
Default: disabled (files will be removed after processing).</p>
</dd></dl>

</div>
<div class="section" id="id42">
<h3>Examples<a class="headerlink" href="#id42" title="Permalink to this headline">¶</a></h3>
<p>Run padding with defaults:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python PadAlignment.py <span class="se">\</span>
  --reference_alignment data/master_alignment.fasta
</pre></div>
</div>
<p>Keep intermediate padded subalignments:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python PadAlignment.py <span class="se">\</span>
  --reference_alignment data/master_alignment.fasta <span class="se">\</span>
  --keep_intermediate_files
</pre></div>
</div>
<p>Use custom directories for input and output:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python PadAlignment.py <span class="se">\</span>
  --reference_alignment data/master_alignment.fasta <span class="se">\</span>
  --input_dir work/Nextalign/query_aln <span class="se">\</span>
  --output_dir results/Pad-alignment
</pre></div>
</div>
</div>
<div class="section" id="id43">
<h3>Notes<a class="headerlink" href="#id43" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>The <strong>reference alignment</strong> is used as a template: its gaps are inserted into
each subalignment so all sequences share the same coordinate space.</p></li>
<li><p>By default, intermediate padded subalignment files are cleaned up automatically.
Use <code class="docutils literal notranslate"><span class="pre">--keep_intermediate_files</span></code> to retain them for debugging.</p></li>
<li><p>After padding, the script also removes redundant sequences to produce a clean
merged alignment.</p></li>
</ul>
</div>
</div>
<div class="section" id="calculate-alignment-coordinates">
<h2>Calculate Alignment Coordinates<a class="headerlink" href="#calculate-alignment-coordinates" title="Permalink to this headline">¶</a></h2>
<p>The script <code class="docutils literal notranslate"><span class="pre">CalcAlignmentCords.py</span></code> calculates genome and CDS coordinates
for sequences based on a padded alignment and a master GFF3 annotation file.
It aligns features to the master reference and outputs a table of coordinates.</p>
<div class="section" id="id44">
<h3>Usage<a class="headerlink" href="#id44" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python CalcAlignmentCords.py --paded_alignment &lt;FASTA&gt; --master_accession &lt;ACC&gt; --master_gff &lt;FILE&gt; <span class="o">[</span>options<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="id45">
<h3>Arguments<a class="headerlink" href="#id45" title="Permalink to this headline">¶</a></h3>
<dl class="option">
<dt id="cmdoption-paded-alignment">
<code class="sig-name descname">-i</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--paded_alignment</code><code class="sig-prename descclassname"> &lt;FILE(S)&gt;</code><a class="headerlink" href="#cmdoption-paded-alignment" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>(Required)</strong> Input FASTA alignment file(s).
Can be a single file or multiple sequence files.</p>
</dd></dl>

<dl class="option">
<dt id="id46">
<code class="sig-name descname">-b</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--tmp_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id46" title="Permalink to this definition">¶</a></dt>
<dd><p>Base working directory.
Default: <code class="docutils literal notranslate"><span class="pre">tmp</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id47">
<code class="sig-name descname">-d</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--output_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id47" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory to store processed data and results.
Default: <code class="docutils literal notranslate"><span class="pre">Tables</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id48">
<code class="sig-name descname">-o</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--output_file</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#id48" title="Permalink to this definition">¶</a></dt>
<dd><p>Output TSV file containing calculated coordinates.
Default: <code class="docutils literal notranslate"><span class="pre">features.tsv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-master-accession">
<code class="sig-name descname">-m</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--master_accession</code><code class="sig-prename descclassname"> &lt;ACC&gt;</code><a class="headerlink" href="#cmdoption-master-accession" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>(Required)</strong> Master reference accession ID.
Example: <code class="docutils literal notranslate"><span class="pre">NC_001542</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-bh">
<code class="sig-name descname">-bh</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--blast_uniq_hits</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-bh" title="Permalink to this definition">¶</a></dt>
<dd><p>BLAST unique hits file used to refine coordinate mapping.
Default: <code class="docutils literal notranslate"><span class="pre">tmp/Blast/query_uniq_tophits.tsv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-master-gff">
<code class="sig-name descname">-g</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--master_gff</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-master-gff" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>(Required)</strong> Master GFF3 annotation file for the reference sequence.</p>
</dd></dl>

</div>
<div class="section" id="id49">
<h3>Examples<a class="headerlink" href="#id49" title="Permalink to this headline">¶</a></h3>
<p>Calculate coordinates with defaults:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python CalcAlignmentCords.py <span class="se">\</span>
  --paded_alignment data/padded_alignment.fa <span class="se">\</span>
  --master_accession NC_001542 <span class="se">\</span>
  --master_gff data/master.gff
</pre></div>
</div>
<p>Specify a custom output location:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python CalcAlignmentCords.py <span class="se">\</span>
  --paded_alignment data/aln.fa <span class="se">\</span>
  --master_accession NC_001542 <span class="se">\</span>
  --master_gff data/master.gff <span class="se">\</span>
  --output_dir results/Tables <span class="se">\</span>
  --output_file coords.tsv
</pre></div>
</div>
<p>Use a custom BLAST hits file:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python CalcAlignmentCords.py <span class="se">\</span>
  --paded_alignment data/aln.fa <span class="se">\</span>
  --master_accession NC_001542 <span class="se">\</span>
  --master_gff data/master.gff <span class="se">\</span>
  --blast_uniq_hits work/Blast/query_uniq_tophits.tsv
</pre></div>
</div>
</div>
<div class="section" id="id50">
<h3>Notes<a class="headerlink" href="#id50" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>The <strong>padded alignment</strong> should be generated by previous steps (e.g., Nextalign + padding).</p></li>
<li><p>The <strong>master GFF3</strong> file provides the annotation used for feature coordinate mapping.</p></li>
<li><p>The <strong>master accession</strong> defines the coordinate reference system.</p></li>
<li><p>The script also leverages BLAST unique hits to refine CDS boundaries where applicable.</p></li>
</ul>
</div>
</div>
<div class="section" id="software-version-checker">
<h2>Software Version Checker<a class="headerlink" href="#software-version-checker" title="Permalink to this headline">¶</a></h2>
<p>The script <code class="docutils literal notranslate"><span class="pre">SoftwareVersion.py</span></code> collects information about available software
versions and generates a TSV summary table. This is useful for reproducibility
and workflow documentation.</p>
<div class="section" id="id51">
<h3>Usage<a class="headerlink" href="#id51" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python SoftwareVersion.py <span class="o">[</span>options<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="id52">
<h3>Arguments<a class="headerlink" href="#id52" title="Permalink to this headline">¶</a></h3>
<dl class="option">
<dt id="id53">
<code class="sig-name descname">-d</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--tmp_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id53" title="Permalink to this definition">¶</a></dt>
<dd><p>Temporary working directory.
Default: <code class="docutils literal notranslate"><span class="pre">tmp</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id54">
<code class="sig-name descname">-o</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--output_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id54" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory where the version summary table will be saved.
Default: <code class="docutils literal notranslate"><span class="pre">Software_info</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-table-name">
<code class="sig-name descname">-f</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--table_name</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-table-name" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of the TSV file to store software version information.
Default: <code class="docutils literal notranslate"><span class="pre">software_info.tsv</span></code></p>
</dd></dl>

</div>
<div class="section" id="id55">
<h3>Examples<a class="headerlink" href="#id55" title="Permalink to this headline">¶</a></h3>
<p>Generate a software version table with defaults:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python SoftwareVersion.py
</pre></div>
</div>
<p>Save the version table with a custom name:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python SoftwareVersion.py <span class="se">\</span>
  --table_name versions.tsv
</pre></div>
</div>
<p>Write results to a custom directory:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python SoftwareVersion.py <span class="se">\</span>
  --output_dir results/Software_info
</pre></div>
</div>
</div>
<div class="section" id="id56">
<h3>Notes<a class="headerlink" href="#id56" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>The output is a TSV file listing the detected software and their versions.</p></li>
<li><p>Useful for documenting pipeline environments and ensuring reproducibility.</p></li>
<li><p>The script can be run at the end of a workflow to capture the environment state.</p></li>
</ul>
</div>
</div>
<div class="section" id="generate-database-ready-tables">
<h2>Generate Database-Ready Tables<a class="headerlink" href="#generate-database-ready-tables" title="Permalink to this headline">¶</a></h2>
<p>The script <code class="docutils literal notranslate"><span class="pre">GenerateTable.py</span></code> produces a set of <strong>TSV tables</strong> ready for
loading into an SQLite database. It consolidates information from the GenBank
matrix, BLAST unique hits, and (padded) multiple-sequence alignments, and can
use Nextalign outputs as needed.</p>
<div class="section" id="id57">
<h3>Usage<a class="headerlink" href="#id57" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python GenerateTable.py <span class="o">[</span>options<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="id58">
<h3>Arguments<a class="headerlink" href="#id58" title="Permalink to this headline">¶</a></h3>
<dl class="option">
<dt id="id59">
<code class="sig-name descname">-g</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--genbank_matrix</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#id59" title="Permalink to this definition">¶</a></dt>
<dd><p>GenBank matrix TSV file.
Default: <code class="docutils literal notranslate"><span class="pre">tmp/GenBank-matrix/gB_matrix_raw.tsv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id60">
<code class="sig-name descname">-b</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--base_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id60" title="Permalink to this definition">¶</a></dt>
<dd><p>Base working directory.
Default: <code class="docutils literal notranslate"><span class="pre">tmp</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id61">
<code class="sig-name descname">-o</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--output_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id61" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory where all DB-ready TSV files are written.
Default: <code class="docutils literal notranslate"><span class="pre">Tables</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-host-taxa">
<code class="sig-name descname">-f</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--host_taxa</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-host-taxa" title="Permalink to this definition">¶</a></dt>
<dd><p>Host taxa lookup table.
Default: <code class="docutils literal notranslate"><span class="pre">host_taxa.tsv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-blast-hits">
<code class="sig-name descname">-bh</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--blast_hits</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-blast-hits" title="Permalink to this definition">¶</a></dt>
<dd><p>BLASTN unique hits table (top hits per query).
Default: <code class="docutils literal notranslate"><span class="pre">tmp/Blast/query_uniq_tophits.tsv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-p">
<code class="sig-name descname">-p</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--paded_aln</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-p" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>Padded</strong> alignment FASTA file (reference-anchored, gap-padded).
Default: <code class="docutils literal notranslate"><span class="pre">tmp/Pad-alignment/NC_001542.aligned_merged_MSA.fasta</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id62">
<code class="sig-name descname">-n</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--nextalign_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id62" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory containing Nextalign alignment outputs.
Default: <code class="docutils literal notranslate"><span class="pre">tmp/Nextalign/</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-email">
<code class="sig-name descname">-e</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--email</code><code class="sig-prename descclassname"> &lt;EMAIL&gt;</code><a class="headerlink" href="#cmdoption-email" title="Permalink to this definition">¶</a></dt>
<dd><p>Contact email (e.g., for NCBI-related requests/logging).
Default: <code class="docutils literal notranslate"><span class="pre">your-email&#64;example.com</span></code></p>
</dd></dl>

</div>
<div class="section" id="id63">
<h3>Examples<a class="headerlink" href="#id63" title="Permalink to this headline">¶</a></h3>
<p>Run with defaults (writes TSVs into <code class="docutils literal notranslate"><span class="pre">Tables/</span></code>):</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python GenerateTable.py
</pre></div>
</div>
<p>Specify custom inputs and output location:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python GenerateTable.py <span class="se">\</span>
  --genbank_matrix data/gB_matrix_raw.tsv <span class="se">\</span>
  --blast_hits results/Blast/query_uniq_tophits.tsv <span class="se">\</span>
  --paded_aln results/Pad-alignment/aligned_merged_MSA.fasta <span class="se">\</span>
  --nextalign_dir results/Nextalign <span class="se">\</span>
  --output_dir results/Tables <span class="se">\</span>
  --email me@lab.org
</pre></div>
</div>
</div>
<div class="section" id="id64">
<h3>Notes<a class="headerlink" href="#id64" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Ensure the <strong>padded alignment</strong> corresponds to the same master reference used
in previous steps, so feature coordinates remain consistent.</p></li>
<li><p>The <strong>BLAST unique hits</strong> file should contain one best hit per query (or as
expected by your pipeline).</p></li>
<li><p>Output TSVs are organized in <code class="docutils literal notranslate"><span class="pre">--output_dir</span></code> and are intended for direct
ingestion into your SQLite schema.</p></li>
</ul>
</div>
</div>
<div class="section" id="create-sqlite-database">
<h2>Create SQLite Database<a class="headerlink" href="#create-sqlite-database" title="Permalink to this headline">¶</a></h2>
<p>The script <code class="docutils literal notranslate"><span class="pre">CreateSqliteTable.py</span></code> creates an <strong>SQLite database</strong> from a
collection of input tables, metadata, alignments, and reference files.
It consolidates processed results into a structured database ready for analysis.</p>
<div class="section" id="id65">
<h3>Usage<a class="headerlink" href="#id65" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python CreateSqliteTable.py <span class="o">[</span>options<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="id66">
<h3>Arguments<a class="headerlink" href="#id66" title="Permalink to this headline">¶</a></h3>
<dl class="option">
<dt id="cmdoption-meta-data">
<code class="sig-name descname">-m</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--meta_data</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-meta-data" title="Permalink to this definition">¶</a></dt>
<dd><p>Metadata table (GenBank matrix TSV).
Default: <code class="docutils literal notranslate"><span class="pre">tmp/GenBank-matrix/gB_matrix_raw.tsv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id67">
<code class="sig-name descname">-b</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--base_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id67" title="Permalink to this definition">¶</a></dt>
<dd><p>Base working directory.
Default: <code class="docutils literal notranslate"><span class="pre">tmp</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id68">
<code class="sig-name descname">-o</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--output_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id68" title="Permalink to this definition">¶</a></dt>
<dd><p>Directory where the SQLite database will be stored.
Default: <code class="docutils literal notranslate"><span class="pre">SqliteDB</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-rf">
<code class="sig-name descname">-rf</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--features</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-rf" title="Permalink to this definition">¶</a></dt>
<dd><p>Features table TSV.
Default: <code class="docutils literal notranslate"><span class="pre">tmp/Tables/features.tsv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-pad-aln">
<code class="sig-name descname">-p</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--pad_aln</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-pad-aln" title="Permalink to this definition">¶</a></dt>
<dd><p>Padded alignment table (TSV).
Default: <code class="docutils literal notranslate"><span class="pre">tmp/Tables/sequence_alignment.tsv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-gene-info">
<code class="sig-name descname">-g</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--gene_info</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-gene-info" title="Permalink to this definition">¶</a></dt>
<dd><p>Gene information table (CSV).
Default: <code class="docutils literal notranslate"><span class="pre">generic/rabv/Tables/gene_info.csv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-mc">
<code class="sig-name descname">-mc</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--m49_countries</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-mc" title="Permalink to this definition">¶</a></dt>
<dd><p>UN M49 countries CSV file.
Default: <code class="docutils literal notranslate"><span class="pre">assets/m49_country.csv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-mir">
<code class="sig-name descname">-mir</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--m49_interm_region</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-mir" title="Permalink to this definition">¶</a></dt>
<dd><p>UN M49 intermediate regions CSV file.
Default: <code class="docutils literal notranslate"><span class="pre">assets/m49_intermediate_region.csv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-mr">
<code class="sig-name descname">-mr</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--m49_regions</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-mr" title="Permalink to this definition">¶</a></dt>
<dd><p>UN M49 regions CSV file.
Default: <code class="docutils literal notranslate"><span class="pre">assets/m49_region.csv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-msr">
<code class="sig-name descname">-msr</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--m49_sub_regions</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-msr" title="Permalink to this definition">¶</a></dt>
<dd><p>UN M49 sub-regions CSV file.
Default: <code class="docutils literal notranslate"><span class="pre">assets/m49_sub_region.csv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-proj-settings">
<code class="sig-name descname">-s</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--proj_settings</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-proj-settings" title="Permalink to this definition">¶</a></dt>
<dd><p>Project/software settings TSV.
Default: <code class="docutils literal notranslate"><span class="pre">tmp/Software_info/software_info.tsv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-fa">
<code class="sig-name descname">-fa</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--fasta_sequences</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-fa" title="Permalink to this definition">¶</a></dt>
<dd><p>FASTA sequence file used in the project.
Default: <code class="docutils literal notranslate"><span class="pre">tmp/GenBank-matrix/sequences.fa</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-insertion-file">
<code class="sig-name descname">-i</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--insertion_file</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-insertion-file" title="Permalink to this definition">¶</a></dt>
<dd><p>Nextalign insertions TSV file.
Default: <code class="docutils literal notranslate"><span class="pre">tmp/Tables/insertions.tsv</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-db-name">
<code class="sig-name descname">-d</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--db_name</code><code class="sig-prename descclassname"> &lt;STR&gt;</code><a class="headerlink" href="#cmdoption-db-name" title="Permalink to this definition">¶</a></dt>
<dd><p>Name of the SQLite database (without <code class="docutils literal notranslate"><span class="pre">.db</span></code> extension).
Default: <code class="docutils literal notranslate"><span class="pre">gdb</span></code></p>
</dd></dl>

</div>
<div class="section" id="id69">
<h3>Examples<a class="headerlink" href="#id69" title="Permalink to this headline">¶</a></h3>
<p>Run with default settings:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python CreateSqliteTable.py
</pre></div>
</div>
<p>Specify a custom database name and output directory:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python CreateSqliteTable.py <span class="se">\</span>
  --db_name rabv_database <span class="se">\</span>
  --output_dir results/SqliteDB
</pre></div>
</div>
<p>Use custom metadata and features tables:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python CreateSqliteTable.py <span class="se">\</span>
  --meta_data data/gB_matrix_raw.tsv <span class="se">\</span>
  --features data/features.tsv <span class="se">\</span>
  --gene_info data/gene_info.csv <span class="se">\</span>
  --db_name custom_gdb
</pre></div>
</div>
</div>
<div class="section" id="id70">
<h3>Notes<a class="headerlink" href="#id70" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>The script integrates <strong>metadata, features, alignments, gene info, country codes, and insertions</strong> into a unified SQLite database.</p></li>
<li><p>Ensure that all required TSV/CSV inputs exist before running.</p></li>
<li><p>The output database will be stored in the directory specified by <code class="docutils literal notranslate"><span class="pre">--output_dir</span></code> with the name <code class="docutils literal notranslate"><span class="pre">&lt;db_name&gt;.db</span></code>.</p></li>
</ul>
</div>
</div>
<div class="section" id="genbank-sequence-submitter">
<h2>GenBank Sequence Submitter<a class="headerlink" href="#genbank-sequence-submitter" title="Permalink to this headline">¶</a></h2>
<p>The script <code class="docutils literal notranslate"><span class="pre">GenBankSequenceSubmitter.py</span></code> prepares sequence data and metadata
for <strong>GenBank submission</strong>. It uses a padded alignment, metadata, a master GFF3,
and a submission template to generate submission-ready files.</p>
<div class="section" id="id71">
<h3>Usage<a class="headerlink" href="#id71" title="Permalink to this headline">¶</a></h3>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python GenBankSequenceSubmitter.py --sequence_dir &lt;DIR&gt; --metadata &lt;FILE&gt; --ncbi_submission_template &lt;FILE&gt; --gff_file &lt;FILE&gt; --vgtk-db &lt;DB&gt; <span class="o">[</span>options<span class="o">]</span>
</pre></div>
</div>
</div>
<div class="section" id="id72">
<h3>Arguments<a class="headerlink" href="#id72" title="Permalink to this headline">¶</a></h3>
<dl class="option">
<dt id="cmdoption-sequence-dir">
<code class="sig-name descname">-q</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--sequence_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#cmdoption-sequence-dir" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>(Required)</strong> Directory containing input FASTA sequence files
(single or multiple).</p>
</dd></dl>

<dl class="option">
<dt id="id73">
<code class="sig-name descname">-t</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--tmp_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id73" title="Permalink to this definition">¶</a></dt>
<dd><p>Temporary working directory.
Default: <code class="docutils literal notranslate"><span class="pre">tmp</span></code></p>
</dd></dl>

<dl class="option">
<dt id="id74">
<code class="sig-name descname">-o</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--output_dir</code><code class="sig-prename descclassname"> &lt;DIR&gt;</code><a class="headerlink" href="#id74" title="Permalink to this definition">¶</a></dt>
<dd><p>Output directory where processed results and submission files are saved.
Default: <code class="docutils literal notranslate"><span class="pre">Table2asn</span></code></p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-metadata">
<code class="sig-name descname">-m</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--metadata</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-metadata" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>(Required)</strong> Metadata file (tab-delimited).
Must include required fields for submission.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-ncbi-submission-template">
<code class="sig-name descname">-n</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--ncbi_submission_template</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-ncbi-submission-template" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>(Required)</strong> NCBI submission template file (generated via:
<cite>https://submit.ncbi.nlm.nih.gov/genbank/template/submission/</cite>).</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-gff">
<code class="sig-name descname">-gff</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--gff_file</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-gff" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>(Required)</strong> Master reference GFF3 annotation file.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-db">
<code class="sig-name descname">-db</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--vgtk-db</code><code class="sig-prename descclassname"> &lt;FILE&gt;</code><a class="headerlink" href="#cmdoption-db" title="Permalink to this definition">¶</a></dt>
<dd><p><strong>(Required)</strong> VGTK SQLite database file.</p>
</dd></dl>

<dl class="option">
<dt id="cmdoption-gp">
<code class="sig-name descname">-gp</code><code class="sig-prename descclassname"></code><code class="sig-prename descclassname">, </code><code class="sig-name descname">--gaps_to_ignore</code><code class="sig-prename descclassname"> &lt;N&gt;</code><a class="headerlink" href="#cmdoption-gp" title="Permalink to this definition">¶</a></dt>
<dd><p>Length of alignment gaps to ignore during processing.
Default: <code class="docutils literal notranslate"><span class="pre">30</span></code></p>
</dd></dl>

</div>
<div class="section" id="id75">
<h3>Examples<a class="headerlink" href="#id75" title="Permalink to this headline">¶</a></h3>
<p>Prepare sequences for submission using defaults:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python GenBankSequenceSubmitter.py <span class="se">\</span>
  --sequence_dir data/sequences <span class="se">\</span>
  --metadata data/metadata.tsv <span class="se">\</span>
  --ncbi_submission_template templates/template.sbt <span class="se">\</span>
  --gff_file reference/master.gff3 <span class="se">\</span>
  --vgtk-db db/vgtk.sqlite
</pre></div>
</div>
<p>Specify a custom gap threshold:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python GenBankSequenceSubmitter.py <span class="se">\</span>
  --sequence_dir data/sequences <span class="se">\</span>
  --metadata data/metadata.tsv <span class="se">\</span>
  --ncbi_submission_template templates/template.sbt <span class="se">\</span>
  --gff_file reference/master.gff3 <span class="se">\</span>
  --vgtk-db db/vgtk.sqlite <span class="se">\</span>
  --gaps_to_ignore <span class="m">50</span>
</pre></div>
</div>
</div>
<div class="section" id="id76">
<h3>Notes<a class="headerlink" href="#id76" title="Permalink to this headline">¶</a></h3>
<ul class="simple">
<li><p>Ensure the <strong>submission template</strong> matches your organism and project type.</p></li>
<li><p>The <strong>VGTK database</strong> provides reference metadata for consistency.</p></li>
<li><p>The generated output in <code class="docutils literal notranslate"><span class="pre">--output_dir</span></code> will include files ready for upload to NCBI’s submission portal.</p></li>
<li><p>Adjust <code class="docutils literal notranslate"><span class="pre">--gaps_to_ignore</span></code> if small insertions/deletions should be tolerated.</p></li>
</ul>
</div>
</div>
</div>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">V-gTK</a></h1>








<h3>Navigation</h3>
<p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="prerequisite.html">Prerequisite</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Functions</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#genbank-xml-downloader">GenBank XML Downloader</a></li>
<li class="toctree-l2"><a class="reference internal" href="#genbank-xml-parser">GenBank XML Parser</a></li>
<li class="toctree-l2"><a class="reference internal" href="#download-gff3-from-ncbi">Download GFF3 from NCBI</a></li>
<li class="toctree-l2"><a class="reference internal" href="#validate-genbank-matrix">Validate GenBank Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="#add-missing-data-to-genbank-matrix">Add Missing Data to GenBank Matrix</a></li>
<li class="toctree-l2"><a class="reference internal" href="#filter-and-extract-sequences">Filter and Extract Sequences</a></li>
<li class="toctree-l2"><a class="reference internal" href="#blast-alignment">BLAST Alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="#nextalign-alignment">Nextalign Alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="#pad-alignment">Pad Alignment</a></li>
<li class="toctree-l2"><a class="reference internal" href="#calculate-alignment-coordinates">Calculate Alignment Coordinates</a></li>
<li class="toctree-l2"><a class="reference internal" href="#software-version-checker">Software Version Checker</a></li>
<li class="toctree-l2"><a class="reference internal" href="#generate-database-ready-tables">Generate Database-Ready Tables</a></li>
<li class="toctree-l2"><a class="reference internal" href="#create-sqlite-database">Create SQLite Database</a></li>
<li class="toctree-l2"><a class="reference internal" href="#genbank-sequence-submitter">GenBank Sequence Submitter</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="rabv-resource.html">Building Rabies(RABV) resource</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="prerequisite.html" title="previous chapter">Prerequisite</a></li>
      <li>Next: <a href="rabv-resource.html" title="next chapter">Building Rabies(RABV) resource</a></li>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2025, Joseph Hughes, Laura Mojsiejczuk, Dana Allen, Sandeep Kasaragod.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 2.4.4</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.12</a>
      
      |
      <a href="_sources/functions.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>